DM-37609: Resurrect prototype client/server and add tests

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Initial test server

timj committed:
Simplify dataset type query

The expression is now a path component rather than query param.

timj committed:
Add method to get all dataset types

timj committed:
Add getDatasetType

timj committed:
Correctly annotate with response_model

timj committed:
Fix some model return types

timj committed:
Add some features from testing

* dimension universe
* return butler config
* query datasets types now properly handles expressions

timj committed:
Use ExpressionQueryParameter model for dataset types query

timj committed:
Experiment with using pydantic for all parameters

timj committed:
Switch query dimension records to data model parameter

timj committed:
Add query data IDs

timj committed:
Use the new kwargs() method on Query

timj committed:
Clean up some collection queries

timj committed:
Remove unused imports

timj committed:
Create butler at launch

This significantly improves performance.

timj committed:
Switch to using a "butler" subapp

This means that the service runs at /butler

timj committed:
Change class for remote registry

timj committed:
Add v1 to REST API

timj committed:
Stop using subapp and add /butler/ explicitly to path

timj committed:
Use enum for CollectionType and use .from_names class method.

timj committed:
Change collection chain interface

timj committed:
Use the proper DatasetId rather than int+UUID

timj committed:
Add a butler instance as an API dependency

timj committed:
Raise HTTPException if an id is not found

timj committed:
Clean up getDatasetLocations in presence of bad ID

timj committed:
Add missing response models

timj committed:
Be explicit about Query parameters

timj committed:
Create read-only and read-write global butlers

timj committed:
Run black and isort on the server prototype

timj committed:
Fix mypy warnings in client/server

timj committed:
Modernize type annotations in client/server

timj committed:
Stop using deepcopy to copy configs

Use the same method we use when copying other mappings.
Deepcopy causes problems if we try to attach objects to a config
after reading the config (for example in client/server testing).

timj committed:
Protect the BUTLER_ROOT replacement against unexpected types

In some edge cases a field that is normally a string will
be replaced by an object that does not support BUTLER_ROOT.
Catch the type error.

timj committed:
Defer reading of Butler repo until we need it

This works around the problem of hardcoding the butler repo location
(which needs to be fixed in the future) for testing.

timj committed:
Add FastAPI to requirements

timj committed:
Add basic client/server registry test

Testing needs to be expanded but this demonstrates that testing
can be done.

timj committed:
Protect test against missing FastAPI

timj committed:
Add test for server queryDimensionRecords

timj committed:
Fix skipping if fastapi is not available

timj committed:
Add remote registry implementations for getDataset and getDatasetType

timj committed:
Improve test coverage for client/server registry

timj committed:
Add remote registry getDatasetLocations

timj committed:
Add test for the get_uri method

This is not going to be a production method but is there
as a demonstration. To make it work we had to implement
refresh in the server but the long term usefulness of this
needs thought. What happens if a dataset type or collection
is added to the server butler? How do the cached butler objects
in the server know they all need to update?

timj committed:
Add client test for getCollectionChain

timj committed:
Fix flake8 warning with enum private method

andy-slac commented on python/lsst/daf/butler/core/repoRelocation.py:
```diff
@@ -64,8 +64,12 @@ def replaceRoot(configRoot: str, butlerRoot: Optional[ResourcePathExpression]) -
+    try:
+        if BUTLER_ROOT_TAG not in configRoot:
+            return configRoot
+    except TypeError:
```
I'm not sure how it can raise TypeError if `configRoot` is a `str` (unless type annotation is wrong).

timj replied:
Funny you should say that. I had to add the try/except because I'm injecting the `TestClient` into the `Config` to override the db field after I have read it and the `Config` classes are not typed at all so eventually what is normally a string ends up being a `TestClient` and the code blows up. Ordinarily the db would be a URL to the server but in the test code we don't run up a server and so this work around has to be used.

andy-slac replied:
Mypy should have complained about it, because `configRoot` is annotated as `str` and that cannot raise TypeError 🙂

andy-slac commented on python/lsst/daf/butler/registries/remote.py:
```diff
@@ -182,7 +218,12 @@ def refresh(self) -> None:
-        pass
+        # How do we know which server should be refreshed?
+        # Should there be caches in the client?
+        response = self._client.put(self._get_url("v1/registry/refresh"))
```
Hmm, it feels a bit odd that server-side refresh should depend on one of the clients asking for it explicitly. 🤔 There could be multiple server instances too, and which one of them receives this request is not known to this client. Maybe the client should only refresh its own caches, and assume that server-side state is always current (I guess that probably means we'd have to disable caching in the Registry on server side).

timj replied:
Yes, my comment in the code there was meant to be me worrying a lot about refresh but in this test code I really need refresh to run in the server because I inject the collections after the initial cached butler is created. You make a later comment on that.

andy-slac commented on python/lsst/daf/butler/registries/remote.py:
```diff
@@ -264,32 +305,55 @@ def removeDatasetType(self, name: str) -> None:
     def getDatasetType(self, name: str) -> DatasetType:
         # Docstring inherited from lsst.daf.butler.registry.Registry
-        raise NotImplementedError()
+        path = f"v1/registry/datasetType/{name}"
```
Maybe avoid hardcoding `v1` in every resource, e.g. do it in `_get_url`?

timj replied:
Yes. Version handling has to be thought about. The `v1` was a suggestion from @rra to make it explicit when we change things in the future. I guess the get_url method could take an optional version override string as well.

andy-slac commented on python/lsst/daf/butler/registries/remote.py:
```diff
@@ -264,32 +305,55 @@ def removeDatasetType(self, name: str) -> None:
+        Parameters
+        ----------
+        dataId : `dict`, `None`, `DataCoordinate`
+            The data ID to normalize.
```
Serialize instead of normalize?

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+
+BUTLER_ROOT = "ci_hsc_gen3/DATA"
+
+log = logging.getLogger("excalibur")
```
⚔️ ?

timj replied:
Two years ago I was having fun with branding and I was fitting in with the SQuaRE naming scheme involving King Arthur names.

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+from lsst.daf.butler.registry import CollectionType
+from pydantic import BaseModel
+
+BUTLER_ROOT = "ci_hsc_gen3/DATA"
```
I guess this is just for testing and will disappear later?

timj replied:
Yes, I override it in the tests anyhow. This will have to come from a configuration file for the real code deployment.

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+    detector: int | None = None
+    instrument: str | None = None
+    physical_filter: str | None = None
+    exposure: int | None = None
```
Is this class used anywhere?

timj replied:
No. It was used but during testing I ended up deleting the code that used it.

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler to use for registry and universe.
+    data_id : `SerializedDatacoordinate` or `None`
```
Coordinate with capital C

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+    return config
+
+
+@app.get("/butler/universe", response_model=dict[str, Any])
```
Why no `v1` for this one?

timj replied:
My feeling was that we would never need to change this and it's a fundamental part of the API that always has to work for all clients (no arguments, returns a dict). Maybe I'm wrong and I should add in a v1 as well here just in case.

timj replied:
Added a v1 here as well.

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+    # refreshed? How do we know the server we are refreshing is used later?
+    # For testing at the moment it is important if a test adds a dataset type
+    # directly in the server since the test client will not see it.
+    butler.registry.refresh()
```
Should it refresh both read-only and writable butlers?

timj replied:
Probably, but I'm having an existential crisis over the refreshing in general. As you say in the overall comment, caching the dataset types and collections in the long-lived server butlers is going to cause a lot of problems since we never really know when something has changed in the database backend that would need a refresh. Globally disabling the cache wold be safest but might be a performance disaster when 10,000 people connect. cc/ @TallJimbo

andy-slac commented on python/lsst/daf/butler/server.py:
```diff
@@ -0,0 +1,441 @@
+
+
+# TimeSpan not yet a pydantic model
+@app.post(
```
Should it be `GET`?

timj replied:
I think it has to be a POST because of the Query parameter in the collections. It's been a while since I had to think about this. If you run the server up it has some very nice automatically created documentation that gets built from knowledge of all the types and models.

andy-slac commented on python/lsst/daf/butler/tests/utils.py:
```diff
@@ -289,4 +289,7 @@ def addDataset(
+        ref = self.butler.put(
+            metric, self.datasetType if datasetType is None else datasetType, dataId, run=run
+        )
+        print("Put dataset: ", ref)
```
Leftover from debugging?

andy-slac commented on tests/test_server.py:
```diff
@@ -0,0 +1,167 @@
+try:
+    from fastapi.testclient import TestClient
+except ImportError:
+    TestClient = None
```
Can you do all optional imports in one `try` block?

timj replied:
Yes, probably, since the failure mode in both cases is not being able to import fastapi and/or httpx.

andy-slac reviewed (approved):
Looks good, few comments. I'm not very familiar yet with httpx/fastapi, will need to learn more. My biggest worry now is server-side state and caching of the registry data in memory, it's clear that we need to think about consistency, probably disable all caching in registry. A related issue is potential caching of responses by proxy servers (if we ever get to that), some of our GET requests are not idempotent, so we may need to disable caching for those too.

timj committed:
Fix two minor typos

timj committed:
Remove unused MaximalDataId model

timj committed:
Add v1 to universe API

It's likely not necessary since this is a simple GET that
returns a dict, but playing it safe won't harm anything.

timj committed:
Factor out v1 string in server API calls

timj committed:
Combine import checks for fastapi

They will all fail anyhow if FastAPI is not installed.

timj committed:
Add news fragment
